File: src/main/scala/tools/jackson/module/scala/ScalaModule.scala
package tools.jackson.module.scala

import tools.jackson.databind.JacksonModule.SetupContext
import tools.jackson.module.scala.deser.{ScalaNumberDeserializersModule, ScalaObjectDeserializerModule, UntypedObjectDeserializerModule}
import tools.jackson.module.scala.introspect.ScalaAnnotationIntrospectorModule

object ScalaModule {

  trait Config {
    /**
     * @return whether the module should support classes built with Scala 3 compiler (default: true)
     */
    def shouldSupportScala3Classes(): Boolean
    /**
     * @return whether the module should support deserializing null collections as empty (default: true)
     */
    def shouldDeserializeNullCollectionsAsEmpty(): Boolean
  }

  class Builder extends Config {
    private val modules = scala.collection.mutable.Buffer[JacksonModule]()
    private var supportScala3Classes = true
    private var deserializeNullCollectionsAsEmpty = true

    def supportScala3Classes(support: Boolean): Builder = {
      supportScala3Classes = support
      this
    }

    def deserializeNullCollectionsAsEmpty(asEmpty: Boolean): Builder = {
      deserializeNullCollectionsAsEmpty = asEmpty
      this
    }

    override def shouldSupportScala3Classes(): Boolean = supportScala3Classes

    override def shouldDeserializeNullCollectionsAsEmpty(): Boolean = deserializeNullCollectionsAsEmpty

    def addModule(module: JacksonModule): Builder = {
      // a module registered twice would contribute its serializers and deserializers twice over
      if (!hasModule(module)) modules.+=(module)
      this
    }

    def removeModule(module: JacksonModule): Builder = {
      val remaining = modules.filterNot(sameModule(_, module)).toList
      modules.clear()
      modules.++=(remaining)
      this
    }

    def hasModule(module: JacksonModule): Boolean = {
      modules.exists(sameModule(_, module))
    }

    // Some builtin modules are registered as an instance of their own, so that a built module keeps
    // state independent of the module object. Matching on the module name as well as on the
    // instance keeps `removeModule(EnumModule)` working for those.
    private def sameModule(existing: JacksonModule, module: JacksonModule): Boolean =
      existing == module || existing.getModuleName == module.getModuleName

    def addAllBuiltinModules(): Builder = {
      addModule(IteratorModule)
      addModule(EnumerationModule)
      addModule(OptionModule)
      addModule(SeqModule)
      addModule(TupleModule)
      addModule(MapModule)
      addModule(SetModule)
      addModule(ScalaNumberDeserializersModule)
      addModule(ScalaAnnotationIntrospectorModule)
      addModule(ScalaObjectDeserializerModule)
      addModule(UntypedObjectDeserializerModule)
      addModule(EitherModule)
      addModule(SymbolModule)
      BuiltinModules.addScalaVersionSpecificModules(this)
      this
    }

    def build(): JacksonModule = {
      val configInstance = this
      new JacksonModule {
        override def getModuleName: String = "ScalaModule"

        override val config: Builder = configInstance
        override def getInitializers(config: Config): Seq[SetupContext => Unit] = {
          modules.toSeq.flatMap(_.getInitializers(config))
        }
      }
    }
  }

  def builder(): Builder = new Builder()

  val defaultBuilder: Config = new Config {
    override def shouldSupportScala3Classes(): Boolean = true
    override def shouldDeserializeNullCollectionsAsEmpty(): Boolean = true
  }
}

File: src/main/java/tools/jackson/module/scala/javadsl/ScalaModule.java
package tools.jackson.module.scala.javadsl;

import tools.jackson.module.scala.*;
import tools.jackson.module.scala.deser.*;
import tools.jackson.module.scala.introspect.ScalaAnnotationIntrospectorModule;

import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;

public final class ScalaModule {
    public static tools.jackson.module.scala.ScalaModule.Builder builder() {
        return ScalaModule$.MODULE$.builder();
    }

    public static EitherModule eitherModule() {
        return EitherModule$.MODULE$;
    }

    public static JacksonModule enumModule() {
        try {
            Class<?> objectClass = Class.forName("tools.jackson.module.scala.EnumModule$");
            VarHandle varHandle = MethodHandles.publicLookup().findStaticVarHandle(
                    objectClass, "MODULE$", objectClass);
            return (JacksonModule) varHandle.get();
        } catch (Throwable t) {
            return null;
        }
    }

    public static EnumerationModule enumerationModule() {
        return EnumerationModule$.MODULE$;
    }

    public static IteratorModule iteratorModule() {
        return IteratorModule$.MODULE$;
    }

    public static IterableModule iterableModule() {
        return IterableModule$.MODULE$;
    }

    public static OptionModule optionModule() {
        return OptionModule$.MODULE$;
    }

    public static TupleModule tupleModule() {
        return TupleModule$.MODULE$;
    }

    public static MapModule mapModule() {
        return MapModule$.MODULE$;
    }

    public static SetModule setModule() {
        return SetModule$.MODULE$;
    }

    public static SymbolModule symbolModule() {
        return SymbolModule$.MODULE$;
    }

    public static ScalaAnnotationIntrospectorModule scalaAnnotationIntrospectorModule() {
        return ScalaAnnotationIntrospectorModule.newStandaloneInstance();
    }

    public static ScalaNumberDeserializersModule scalaNumberDeserializersModule() {
        return ScalaNumberDeserializersModule$.MODULE$;
    }

    public static ScalaObjectDeserializerModule scalaObjectDeserializerModule() {
        return ScalaObjectDeserializerModule$.MODULE$;
    }

    public static UntypedObjectDeserializerModule untypedObjectDeserializerModule() {
        return UntypedObjectDeserializerModule$.MODULE$;
    }
}
